optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java

```diff
@@ -0,0 +1,407 @@
+        if (stepScope.getStep() != null) {
+            // Increase stepIndex by 1, because it's a preliminary action
+            ApplyStepOperation<Solution_> stepOperation = new ApplyStepOperation<>(
+                    stepIndex + 1, stepScope.getStep(), stepScope.getScore());
+            for (int i = 0; i < moveThreadCount; i++) {
+                operationQueue.add(stepOperation);
+            }
+        }
+        // TODO latch barrier
+    }
+
+    private static abstract class MoveThreadOperation<Solution_> {
+
+        @Override
+        public String toString() {
+            return getClass().getSimpleName();
+        }
+
+    }
+
+    private static class SetupOperation<Solution_> extends MoveThreadOperation<Solution_> {
+
+        private final InnerScoreDirector<Solution_> innerScoreDirector;
+
+        public SetupOperation(InnerScoreDirector<Solution_> innerScoreDirector) {
+            this.innerScoreDirector = innerScoreDirector;
+        }
+
+        public InnerScoreDirector<Solution_> getScoreDirector() {
+            return innerScoreDirector;
+        }
+    }
+
+    private static class DestroyOperation<Solution_> extends MoveThreadOperation<Solution_> {
+
+    }
+
+    private static class ApplyStepOperation<Solution_> extends MoveThreadOperation<Solution_> {
+
+        private final int stepIndex;
+        private final Move<Solution_> step;
+        private final Score score;
+
+        public ApplyStepOperation(int stepIndex, Move<Solution_> step, Score score) {
+            this.stepIndex = stepIndex;
+            this.step = step;
+            this.score = score;
+        }
+
+        public int getStepIndex() {
+            return stepIndex;
+        }
+
+        public Move<Solution_> getStep() {
+            return step;
+        }
+
+        public Score getScore() {
+            return score;
+        }
+    }
+
+    private static class MoveEvaluationOperation<Solution_> extends MoveThreadOperation<Solution_> {
+
+        private final int stepIndex;
+        private final int moveIndex;
+        private final Move<Solution_> move;
+
+        public MoveEvaluationOperation(int stepIndex, int moveIndex, Move<Solution_> move) {
+            this.stepIndex = stepIndex;
+            this.moveIndex = moveIndex;
+            this.move = move;
+        }
+
+        public int getStepIndex() {
+            return stepIndex;
+        }
+
+        public int getMoveIndex() {
+            return moveIndex;
+        }
+
+        public Move<Solution_> getMove() {
+            return move;
+        }
+    }
+
+    private class MoveThreadRunner implements Runnable {
+
+        private int moveThreadIndex;
+        private InnerScoreDirector<Solution_> scoreDirector = null;
+
+        public MoveThreadRunner(int moveThreadIndex) {
+            this.moveThreadIndex = moveThreadIndex;
+        }
+
+        @Override
+        public void run() {
```

Comment:
Too long and complicated method, breaking it into private (or package private) methods would make it more readable and testable.